### PR TITLE
[SPARK-24046][SS] Fix rate source when rowsPerSecond <= rampUpTime  

### PR DESCRIPTION
**Pull request by maasg**
## What changes were proposed in this pull request?

Fixes the ramp-up of the rate source for the case `rowsPerSecond <= rampUpTime` which previously resulted in a flat ramp-up (0-value until the clock reached `rampUpTime`. For example: 
image-2018-04-22-22-06-49-202

See details in  [SPARK-24046](https://issues.apache.org/jira/browse/SPARK-24046)

This implementation replaces the original code. Uses the distance quadratic function, meaning that ramp-up is a smooth parabole that joins the linear constant rate at `rampUpTime`:
Original vs. this PR:
ramp-up-break

## How was this patch tested?

- Improved and extended existing unit tests.
- Validated the algorithm visually in a notebook for many different scenarios

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProvider.scala [hunk 1]
```diff
@@ -101,25 +101,10 @@ object RateStreamProvider {
...
-      }
-    } else {
-      // rampUpPart is just a special case of the above formula: rampUpTimeSeconds == seconds
-      val rampUpPart = valueAtSecond(rampUpTimeSeconds, rowsPerSecond, rampUpTimeSeconds)
-      rampUpPart + (seconds - rampUpTimeSeconds) * rowsPerSecond
-    }
+    val delta = rowsPerSecond.toDouble / rampUpTimeSeconds
+    val rampUpSeconds = if (seconds <= rampUpTimeSeconds) seconds else rampUpTimeSeconds
+    val afterRampUpSeconds = if (seconds > rampUpTimeSeconds ) seconds - rampUpTimeSeconds else 0
+    // Use classic distance formula based on accelaration: ut + ½at2
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProvider.scala]: nit: acceleration

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProvider.scala [hunk 2]
```diff
@@ -101,25 +101,10 @@ object RateStreamProvider {
...
+    Math.floor(rampUpSeconds * rampUpSeconds * delta / 2).toLong + afterRampUpSeconds * rowsPerSecond
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProvider.scala]: nit: >100 characters

**Review by xuanyuanking — COMMENTED** [review]
Expect some nits, the ut changes confuse me, it seems a clever resolution but I think it should not change the original semantics of RateStreamProvider.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProviderSuite.scala [hunk 3]
```diff
@@ -173,55 +173,154 @@ class RateSourceSuite extends StreamTest {
...
+      rps <- rowsPerSec
+      rampUp <- rampUpSec
+    } yield {
+      assert(valueAtSecond(seconds = 0, rowsPerSecond = rps, rampUpTimeSeconds = rampUp) === 0)
+    }
 
-    assert(valueAtSecond(seconds = 0, rowsPerSecond = 5, rampUpTimeSeconds = 2) === 0)
-    assert(valueAtSecond(seconds = 1, rowsPerSecond = 5, rampUpTimeSeconds = 2) === 1)
-    assert(valueAtSecond(seconds = 2, rowsPerSecond = 5, rampUpTimeSeconds = 2) === 3)
-    assert(valueAtSecond(seconds = 3, rowsPerSecond = 5, rampUpTimeSeconds = 2) === 8)
```
Comment (xuanyuanking) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProviderSuite.scala]: I try your implement local and it changes the original behavior
```
valueAtSecond(seconds = 1, rowsPerSecond = 5, rampUpTimeSeconds = 2) = 1
valueAtSecond(seconds = 2, rowsPerSecond = 5, rampUpTimeSeconds = 2) = 5
valueAtSecond(seconds = 3, rowsPerSecond = 5, rampUpTimeSeconds = 2) = 10
valueAtSecond(seconds = 4, rowsPerSecond = 5, rampUpTimeSeconds = 2) = 15
```
I think the bug fix should not change the value on `seconds > rampUpTimeSeconds`, just my opinion, you can ping other committers to review.

Comment (maasg) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProviderSuite.scala]: I would agree with that in general, but in this case, the fix is also improving the behavior of the previously working scenario. Also, the numbers in the test seem arbitrary to me. 
Here I compare the current implementation with the proposed solution for case where `rowsPerSecond > rampUpSeconds`:
```scala
for rampUpSeconds = 10 to 100 by 10
valueAtSecond(i, rowsPerSecond = 100, rampUpTimeSeconds=x)
```
Original implementation:
image
Proposed solution:
image

**Comment by maasg** [comment]
pinging @zsxwing:  as the original author of the `RateSourceProvider`, could you review this PR?

**Comment by maasg** [comment]
@holdenk as we discussed in Strata, it would be great if you could give me your opinion on the approach taken in this PR.

**Comment by skonto** [comment]
@felixcheung Could you review pls ?

**Comment by zsxwing** [comment]
This is just a feature to warm up JVMs. Why do we need a smooth increasing for this?

**Comment by maasg** [comment]
@zsxwing Thanks for dropping by. This patch is about fixing the rate ramp up when `rowsPerSecond <= rampUpTime`, which makes the Rate Source produce no data until `rampUpTime` (See  [SPARK-24046](https://issues.apache.org/jira/browse/SPARK-24046)).

The review discussion in this PR is that, while fixing this issue, I introduced a new way of calculating the `rampUp` that makes the previously working scenario of `rowsPerSecond > rampUpTime` smoother and more consistent (as shown in the charts above). 
The original tests verified the ramp-up against some hard-coded values that are changed by the new formula. While the semantics of the 'ramp up' behavior are preserved, the intermediate ramp up values produced are different, which is evidenced in the test. 

I believe the overall code approach is an improvement over the original and the behavior it shows is what we would expect from the description of the 'ramp up'  feature.

What do you think?  Could you review the code changes?

**Comment by gaborgsomogyi** [comment]
I don't see the value to make the ramp-up smooth. As I see it would make unit tests even slower.

**Comment by felixcheung** [comment]
@zsxwing @jose-torres

**Pull request closed without merging**